### PR TITLE
ARROW-14844 [R] Implement `decimal256()`

### PR DESCRIPTION
**Pull request by dragosmg**

### CODE REVIEW

**Review by romainfrancois — COMMENTED** [review]
in addition: https://github.com/dragosmg/arrow/pull/1

**Comment by romainfrancois** [comment]
As with `decimal128` this would not allow for creating arrays directly: 

``` r
library(arrow, warn.conflicts = FALSE)
#> See arrow_info() for available features

Array$create(c(1, 2), type = decimal(4, 3))
#> Error: NotImplemented: Extend
```

<sup>Created on 2021-12-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

The error comes from this class in `r_to_arrow.cpp`

```cpp
template <typename T>
class RPrimitiveConverter<T, enable_if_t<is_decimal_type<T>::value>>
    : public PrimitiveConverter<T, RConverter> {
 public:
  Status Extend(SEXP x, int64_t size, int64_t offset = 0) override {
    return Status::NotImplemented("Extend");
  }
};
```

Maybe we could catch the `NotImplemented` error, then convert to a guessed type, and then cast.

**Comment by romainfrancois** [comment]
Perhaps we could have something like this on the R side: 

```r
Array$create <- function(x, type = NULL) {
  if (!is.null(type)) {
    type <- as_type(type)
  }
  if (inherits(x, "Scalar")) {
    out <- x$as_array()
    if (!is.null(type)) {
      out <- out$cast(type)
    }
    return(out)
  }
  tryCatch(
    vec_to_Array(x, type),
    error = function(cnd) {
      if (!is.null(type)) {
        # try again and then cast
        vec_to_Array(x, NULL)$cast(type)
      } else {
        signalCondition(cnd)
      }
    }
  )
}
```

So that: 

``` r
library(arrow, warn.conflicts = FALSE)
#> See arrow_info() for available features

Array$create(c(1, 2), type = decimal(4, 3))
#> Array
#> <decimal128(4, 3)>
#> [
#>   1.000,
#>   2.000
#> ]
```

<sup>Created on 2021-12-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

File: r/R/type.R [hunk 1]
```diff
@@ -387,6 +389,22 @@ decimal <- function(precision, scale) {
   Decimal128Type__initialize(precision, scale)
 }
 
+#' @rdname data-type
+#' @export
+decimal256 <- function(precision, scale) {
```
[the remark this answers is not in the document]
Comment (dragosmg) [reply r/R/type.R:399]: Great idea! I will implement it.

**Comment by dragosmg** [comment]
This is outdated. Work on `decimal256()` is taking place in a different PR - https://github.com/apache/arrow/pull/11898

**Pull request closed without merging**